pkg/sessiontxn/isolation/base.go

```diff
@@ -468,6 +446,159 @@ func (p *baseTxnContextProvider) getSnapshotByTS(snapshotTS uint64) (kv.Snapshot
+	}
+
+	if txn.IsPipelined() {
+		txn.SetOption(kv.RequestSourceType, "p-dml")
+	} else if tp := p.sctx.GetSessionVars().RequestSourceType; tp != "" {
+		txn.SetOption(kv.RequestSourceType, tp)
+	}
+
+	if sessVars.LoadBasedReplicaReadThreshold > 0 {
+		txn.SetOption(kv.LoadBasedReplicaReadThreshold, sessVars.LoadBasedReplicaReadThreshold)
+	}
+
+	txn.SetOption(kv.CommitHook, func(info string, _ error) { sessVars.LastTxnInfo = info })
+	txn.SetOption(kv.EnableAsyncCommit, sessVars.EnableAsyncCommit)
+	txn.SetOption(kv.Enable1PC, sessVars.Enable1PC)
+	if sessVars.DiskFullOpt != kvrpcpb.DiskFullOpt_NotAllowedOnFull {
+		txn.SetDiskFullOpt(sessVars.DiskFullOpt)
+	}
+	txn.SetOption(kv.InfoSchema, sessVars.TxnCtx.InfoSchema)
+	txn.SetOption(kv.ResourceGroupTagger, sessVars.StmtCtx.GetResourceGroupTagger())
+	txn.SetOption(kv.ExplicitRequestSourceType, sessVars.ExplicitRequestSourceType)
+
+	if p.causalConsistencyOnly || !sessVars.GuaranteeLinearizability {
+		// priority of the sysvar is lower than `start transaction with causal consistency only`
+		txn.SetOption(kv.GuaranteeLinearizability, false)
+	} else {
+		// We needn't ask the TiKV client to guarantee linearizability for auto-commit transactions
+		// because the property is naturally holds:
+		// We guarantee the commitTS of any transaction must not exceed the next timestamp from the TSO.
+		// An auto-commit transaction fetches its startTS from the TSO so its commitTS > its startTS > the commitTS
+		// of any previously committed transactions.
+		// Additionally, it's required to guarantee linearizability for snapshot read-only transactions though
+		// it does take effects on read-only transactions now.
+		txn.SetOption(kv.GuaranteeLinearizability,
+			!sessVars.IsAutocommit() ||
+				sessVars.SnapshotTS > 0 ||
+				p.enterNewTxnType == sessiontxn.EnterNewTxnDefault ||
+				p.enterNewTxnType == sessiontxn.EnterNewTxnWithBeginStmt)
+	}
+}
+
+func (p *baseTxnContextProvider) SetOptionsBeforeCommit(txn kv.Transaction, commitTSChecker func(uint64) bool) error {
+	sessVars := p.sctx.GetSessionVars()
+	// Pipelined dml txn already flushed mutations into stores, so we don't need to set options for them.
+	// Instead, some invariants must be checked to avoid anomalies though are unreachable in designed usages.
+	if p.txn.IsPipelined() {
+		if p.txn.IsPipelined() && !sessVars.TxnCtx.EnableMDL {
+			return errors.New("cannot commit pipelined transaction without Metadata Lock: MDL is OFF")
+		}
+		if len(sessVars.TxnCtx.TemporaryTables) > 0 {
+			return errors.New("pipelined dml with temporary tables is not allowed")
+		}
+		if sessVars.BinlogClient != nil {
+			return errors.New("pipelined dml with binlog is not allowed")
+		}
+		if sessVars.CDCWriteSource != 0 {
+			return errors.New("pipelined dml with CDC source is not allowed")
+		}
+		if commitTSChecker != nil {
+			return errors.New("pipelined dml with commitTS checker is not allowed")
+		}
+		return nil
+	}
+
```

Comment:
Do we forget to set `kv.CommitTSUpperBoundCheck` when `commitTSChecker` is not nil?

Reply:
Added.